ManagementPacks/Community.DataOnDemand/Scripts/Watch-TcpConnections.ps1

```diff
@@ -0,0 +1,539 @@
+        private enum TCP_TABLE_CLASS
+        {
+            TCP_TABLE_BASIC_LISTENER,
+            TCP_TABLE_BASIC_CONNECTIONS,
+            TCP_TABLE_BASIC_ALL,
+            TCP_TABLE_OWNER_PID_LISTENER,
+            TCP_TABLE_OWNER_PID_CONNECTIONS,
+            TCP_TABLE_OWNER_PID_ALL,
+            TCP_TABLE_OWNER_MODULE_LISTENER,
+            TCP_TABLE_OWNER_MODULE_CONNECTIONS,
+            TCP_TABLE_OWNER_MODULE_ALL
+        }
+
+        // http://msdn.microsoft.com/en-us/library/aa366896.aspx
+        private enum MIB_TCP_STATE
+        {
+            MIB_TCP_STATE_CLOSED,
+            MIB_TCP_STATE_LISTEN,
+            MIB_TCP_STATE_SYN_SENT,
+            MIB_TCP_STATE_SYN_RCVD,
+            MIB_TCP_STATE_ESTAB,
+            MIB_TCP_STATE_FIN_WAIT1,
+            MIB_TCP_STATE_FIN_WAIT2,
+            MIB_TCP_STATE_CLOSE_WAIT,
+            MIB_TCP_STATE_CLOSING,
+            MIB_TCP_STATE_LAST_ACK,
+            MIB_TCP_STATE_TIME_WAIT,
+            MIB_TCP_STATE_DELETE_TCB
+        }
+        #endregion
+
+        #region APIs
+        // http://msdn2.microsoft.com/en-us/library/aa366073.aspx
+        private static class IPHelperAPI
+        {
+            [DllImport("iphlpapi.dll", SetLastError = true)]
+            internal static extern uint GetExtendedTcpTable(
+                IntPtr tcpTable,
+                ref int tcpTableLength,
+                bool sort,
+                int ipVersion,
+                TCP_TABLE_CLASS tcpTableType,
+                int reserved);
+        }
+        #endregion
+
+        #region wrapper class
+        private class IPHelperWrapper : IDisposable
+        {
+
+            public const int AF_INET = 2;    // IP_v4 = System.Net.Sockets.AddressFamily.InterNetwork
+            public const int AF_INET6 = 23;  // IP_v6 = System.Net.Sockets.AddressFamily.InterNetworkV6
+
+            // Creates a new wrapper for the local machine
+            public IPHelperWrapper() { }
+
+            // Disposes of this wrapper
+            public void Dispose() { GC.SuppressFinalize(this); }
+
+            public List<MIB_TCPROW_OWNER_PID> GetAllTCPv4Connections()
+            {
+                return GetTCPConnections<MIB_TCPROW_OWNER_PID, MIB_TCPTABLE_OWNER_PID>(AF_INET);
+            }
+
+            public List<MIB_TCP6ROW_OWNER_PID> GetAllTCPv6Connections()
+            {
+                return GetTCPConnections<MIB_TCP6ROW_OWNER_PID, MIB_TCP6TABLE_OWNER_PID>(AF_INET6);
+            }
+
+            public List<IPR> GetTCPConnections<IPR, IPT>(int ipVersion)
+            { //IPR = Row Type, IPT = Table Type
+
+                IPR[] tableRows;
+                int buffSize = 0;
+                FieldInfo dwNumEntriesField = typeof(IPT).GetField("dwNumEntries");
+
+                // how much memory do we need?
+                uint ret = IPHelperAPI.GetExtendedTcpTable(IntPtr.Zero, ref buffSize, true, ipVersion, TCP_TABLE_CLASS.TCP_TABLE_OWNER_PID_ALL, 0);
```

Comment:
For a monitoring task running on the various agent(s) currently of interest, I wanted a low-impact solution, so firing up a new process every second seemed a bad idea. We can certainly look at phasing out other uses of netstat, but that's not a top priority right now.